apps/04TowersOfHanoi.js

```diff
@@ -13,30 +13,92 @@ var stacks = {
   c: []
 };
 
+var counter = 0;
+
 function printStacks() {
   console.log("a: " + stacks.a);
   console.log("b: " + stacks.b);
   console.log("c: " + stacks.c);
 }
 
-function movePiece() {
-  // Your code here
+// starts a new game when won.
+function newGame() {
+  stacks = {
+    a: [4, 3, 2, 1],
+    b: [],
+    c: []
+  };
+  //resets the amount of moves the user makes in the new game.
+  counter = 0;
+  return stacks;
+}
 
+//Moving a piece from one stack to another
+function movePiece(startStack, endStack) {
+  var move = stacks[startStack].pop();
+  stacks[endStack].push(move);
+  //number of moves increases by 1
+  counter++;
+  return true;
 }
 
-function isLegal() {
-  // Your code here
 
+function isLegal(startStack, endStack) {
+  var startLength = stacks[startStack].length;
+  var endLength = stacks[endStack].length;
+  //if the end stack is empty, any piece can go there.
+  if(stacks[endStack].length < 1) {
+    return true;
+  //if the piece being moved is smaller than the piece on the end stack, it is a legal move.
+  } else if((stacks[startStack][startLength - 1]) < (stacks[endStack][endLength - 1])) {
+      if(startStack !== 'a' || startStack !== 'b' || startStack !== 'c') {
+        console.log("Invalid input. Try again.");
+        return false;
+      } else if(startStack !== isNaN) {
+        console.log("Invalid input. Try again.");
+        return false;
+      } 
+
+      if(endStack !== 'a' || endStack !== 'b' || endStack !== 'c' ) {
+        console.log("Invalid input. Try again.");
+        return false;
+      } else if(endStack !== isNaN) {
+        console.log("Invalid input. Try again.");
+        return false;
+      }
+
+      return true;
+  //otherwise illegal
+  } else {
+    console.log("Not a valid move. Try again!");
+    return false;
+  }
 }
 
+//checks if the user has won.
 function checkForWin() {
-  // Your code here
+  if(stacks['b'].length === 4 || stacks['c'].length === 4) {
+    console.log("You Won!");
+    newGame();
+    return true;
+  } else {
+    return false;
+  }
 
 }
 
 function towersOfHanoi(startStack, endStack) {
-  // Your code here
 
+  startStack = startStack.toLowerCase();
+  endStack = endStack.toLowerCase();
+//if the move is legal, move the piece
+  if(isLegal(startStack, endStack) === true) {
```

Comment:
You can pass in the function directly and the `if` will only run if the function returns true.

example: `if(returnTrueOrFalse()) { return "I am true!" }`